[클린코드 리액트 2기 김동호] 페이먼츠 미션 Step 1

psy082 opened the pull request:
안녕하세요 현수님! 
리액트를 안쓰다가 쓰려니깐 힘든 와중에 과제까지 너무 어려워서 오래 걸렸는데 아직도 다 못했다는 변명으로 PR을 시작합니다...!

현재 카드 정보 입력을 위한 컴포넌트만 작성되었고 아직 페이지들은 만들지 못했습니다.
작성한 컴포넌트를 사용하여서 페이지 작성도 빠르게 진행하려고 합니다!

우선 컴포넌트 컨셉을 다음과 같이 정했습니다.
 
1. 카드 정보 입력을 위한 Input의 종류가 확장되더라도 공통적으로 제공되어야 하는
focus 이동 기능, 입력 값 길이 계산 기능 들이 외부에 노출되지 않고 컴포넌트 내부에서 제공되도록
context api, useRef를 활용한 합성 컴포넌트로 작성한다.

OOP 디자인 패턴의 템플릿 메서드 패턴의 컨셉을 차용하려고 했는데요(어쩌면 전략 패턴인 것 같기도 하지만요), 결과적으로 사용자가 Input 컴포넌트에 max length와 validate를 위한 regex를 추가하는 방식으로 커스텀 Input 컴포넌트가 확장되는 걸 의도했습니다. 물론 확장 시에 필요한 정보가 위의 두개 정보만인 상태가 유지된다는 전제 하에서 성립되겠지만요... 

2. 카드 번호, 만료일 등 한 단위의 카드 정보를 입력받는 Container 내부는 ref의 사용으로 uncontrolled component이지만,
Container는 props로 onChange 함수를 노출하기 때문에
Container를 사용하는 외부에서는 Container를 Controlled component로 사용할 수 있도록 작성한다.

Container 내부를 uncontrolled component 로 만든 것은 의도한 바는 아니지만 focus 이동 기능을 제공하기 위해 ref를 사용한 결과로 부차적으로 만들어졌습니다. 

질문

1. 처음에는 기능만 생각하고 컴포넌트 구현을 하다가 나중에는 스타일 관련 정보도 컴포넌트 사용자에게 입력 받아야 한다는 사실을 깨달았습니다. 지금은 스타일을 전부 외부에서 입력 받도록 노출하지는 않았고 width를 위한 %값과 background 사용 여부만 노출했는데, 개발 과정에서 기능보다는 스타일 때문에 컴포넌트의 props가 자주 변경될 여지가 많을 것 같다는 것을 느꼈습니다. 리액트에서는 스타일 관련 변화율을 어떻게 관리하는 지 궁금합니다! 아예 외부에서 style을 지정할 수 있도록 style을 props로 노출하는 것이 더 나을까요? 

2. 지금의 컴포넌트가 css class 이름에 props가 강하게 의존하고 있는 것도 문제로 느끼는 데요, 어쩌면 css in js가 그 모든 정보를 컴포넌트 내부로 감추면서 문제를 해결할 수도 있을 것 같지만, 최근에는 css in js를 사용하지 않다는 글도 읽었던 것 같아서요. 컴포넌트에서 스타일 정보와의 의존 문제를 어떤 관점에서 보는 게 좋을 지도 궁금합니다. 저는 아직 무엇을 기준으로 판단하는 게 좋을지 모르겠습니다...


## 📝 Requirements

### 필수 요구사항

- [x] Storybook 상호 작용 테스트
- [x] 재사용 가능한 Component 작성

### 카드 추가

- [x] `<` (뒤로가기) 버튼 클릭 시, 카드 목록 페이지로 이동한다.
- [x] 카드 번호를 입력 받을 수 있다.
  - [x] 카드 번호는 숫자만 입력가능하다.
  - [x] 카드 번호 4자리마다 `-`가 삽입된다.
  - [ ] 카드 번호는 실시간으로 카드 UI에 반영된다.
  - [x] 카드 번호는 앞 8자리만 숫자로 보여지고, 나머지 숫자는 `*`로 보여진다.
- [x] 만료일을 입력 받을 수 있다.
  - [x] MM / YY 로 placeholder를 적용한다.
  - [x] 월, 년 사이에 자동으로 /가 삽입된다.
  - [ ] 만료일은 실시간으로 카드 UI에 반영된다.
  - [x] 월은 1이상 12이하 숫자여야 한다.
- [x] 보안코드를 입력 받을 수 있다.
  - [x] ~~보안코드는 `*`으로 보여진다.~~ (type="password"를 사용하여 `*` 표시 사용하지 않음)
  - [x] 보안코드는 숫자만 입력가능하다.
- [x] 카드 비밀번호의 앞 2자리를 입력 받을 수 있다.
  - [x] 카드 비밀번호는 각 폼마다 한자리 숫자만 입력가능하다.
  - [x] ~~카드 번호 입력 시, `*`으로 보여진다.~~
- [x] 카드 소유자 이름을 입력 받을 수 있다.
  - [x] 이름은 30자리까지 입력할 수 있다.
  - [x] 이름 입력 폼 위에, 현재 입력 자릿수와 최대 입력 자릿수를 실시간으로 보여준다.
- [ ] (추가) 카드 값을 localStorage에 저장한다
- [ ] (추가) 카드 색상 + 카드 이름을 변경할 수 있다
- [ ] 카드 추가 완료시 카드 등록 완료 페이지로 이동한다.
  - [ ] localStorage 값으로 카드 등록 완료 페이지에서 입력한 값을 확인한다.

psy082 committed:
init: 프로젝트 초기 설정

psy082 committed:
chore: storybook 추가

psy082 committed:
chores: 라이브러리 추가 및 설정

- emotion 설치
- eslint 설치 및 설정

psy082 committed:
feat: card component 구현

psy082 committed:
feat: 카드 정보 입력을 위한 컴포넌트 작성

카드 정보 입력을 위한 Input의 종류가 확장되더라도 공통적으로 제공되어야 하는
focus 이동 기능, 입력 값 길이 계산 기능 들이 외부에 노출되지 않고 컴포넌트 내부에서 제공되도록
context api, useRef를 활용한 합성 컴포넌트로 작성함.

카드 번호, 만료일 등 한 단위의 카드 정보를 입력받는 Container 내부는 ref의 사용으로 uncontrolled component이지만,
Container는 props로 onChange 함수를 노출하기 때문에
Container를 사용하는 외부에서는 Container를 Controlled component로 사용할 수 있도록 작성함.

junghyeonsu commented on package.json:
```diff
@@ -0,0 +1,65 @@
+  "private": true,
+  "dependencies": {
+    "@emotion/react": "^11.10.6",
+    "@testing-library/jest-dom": "^5.16.5",
+    "@testing-library/react": "^13.4.0",
+    "@testing-library/user-event": "^13.5.0",
+    "@types/jest": "^27.5.2",
+    "@types/node": "^16.18.12",
+    "@types/react": "^18.0.28",
+    "@types/react-dom": "^18.0.11",
```
해당 라이브러리들은 `devDependencies`에 있어도 괜찮을 것 같네요~

psy082 replied:
아하 의존성 부분은 cra에서 설정한 걸 그대로 놔두었는데, 그냥 넘겼던 것 같습니다.
그래도 package.json도 문서이니 만큼 옮겨서 작성해보겠습니다 :)

junghyeonsu commented on package.json:
```diff
@@ -0,0 +1,65 @@
+    "eslint-config-prettier": "^8.6.0",
+    "eslint-config-standard-with-typescript": "^34.0.0",
+    "eslint-plugin-import": "^2.27.5",
+    "eslint-plugin-n": "^15.6.1",
```
`eslint-plugin-n` 플러그인을 처음보네요! node를 위한 플러그인이군요.

제가 좋아하는 플러그인도 소개해드릴게요!
[eslint-plugin-json-format](https://www.npmjs.com/package/eslint-plugin-json-format)
json도 포맷팅해주는 플러그인이에요!

[eslint-plugin-simple-import-sort](https://www.npmjs.com/package/eslint-plugin-simple-import-sort)는 import 순서를 포맷팅해주는 플러그인 인데요, 써드파티 라이브러리를 맨 위로 그리고 자신이 작성한 컴포넌트나 모듈같은 경우는 그 아래로 정렬을 자동으로 시켜줘요.

psy082 replied:
`eslint-plugin-n` 플러그인은 다른 플러그인들과 함께 적용된 것 같습니다. 저도 어떤 플러그인이지 확인을 안해봤네요 ㅠ 하나씩 확인을 해봐야 겠습니다. 말씀해 주신 플러그인들은 정말 편리하네용! 추천 감사합니다 👍

junghyeonsu commented on src/Components/Card/Card.tsx:
```diff
@@ -0,0 +1,70 @@
+  };
+}
+
+function Card(props: Props) {
+  const {
+    company,
+    number,
+    owner,
+    expiration: { month, year },
+  } = props;
```
```suggestion
function Card({ company, number, owner, expiration: { month, year }}: Props) {
```

이런식으로도 할 수 있겠네요!

psy082 replied:
앗 맞습니다!

junghyeonsu commented on src/Components/Card/Card.tsx:
```diff
@@ -0,0 +1,70 @@
+  );
+}
+
+const splitCardNumber = (number: string): string[] => {
+  let digits = '';
+  const splittedNumberArr: string[] = [];
+  for (let i = 0; i < number.length; i++) {
+    if (i > 0 && i % 4 === 0) {
+      splittedNumberArr.push(digits);
+      digits = '';
+    }
+    digits += number[i];
+  }
+  splittedNumberArr.push(digits);
+  return splittedNumberArr;
+};
```
추후에는 `splitCardNumber` 이라는 이름말고 `splitNumber`로 만들어서 유틸함수로도 만들 수 있겠네요.
지금은 card라는 도메인에 엮여있는 것 같은데 로직적으로는 크게 card에만 묶여있을 필요는 없는 유틸 함수 같아서요ㅎㅎ

그리고 지금은 4라는 숫자가 매직처럼 느껴지네요. 처음 읽는 사람은 4가 뭔지 모를수도 있겠어요!

`truncateStringToArray` 라는 이름으로 짓고 string과 size를 인자로 받는 유틸을 만들면 되지 않을까요!?

```ts
const truncateStringToArray = (str: string, size: number): string[] => {
  const result: string[] = [];

  for (let i = 0; i < str.length; i += size) {
    const chunk = str.substr(i, size);
    result.push(chunk);
  }

  return result;
};
```

대충 작성해봤어요ㅎㅎ

psy082 replied:
확실히 알려주신 구현 방법이 더 좋은 것 같습니다! 
다만... 입력이 실시간으로 카드에 반영되기 위해서는 카드 번호를 아예 스트링에서 배열로 받아야 한다는 것을 깨달아서 해당 함수는 삭제될 것 같습니다.

코멘트 해주신 부분에 대해 좀 더 이야기 한다면

처음에는 4를 SplittedCardNumber라는 변수에 담으려고 하다가 4개 단위로 잘린 카드 번호 자체가 카드에 대한 도메인적 지식을 담고 있는 것 같아서 SplittedCardNumbe 너무 일반화된 이름인 것 같다는 생각이 들었습니다. 다른 이름을 고민하던 차에 함수가 외부에 노출되지 않았고 카드 컴포넌트와 함께 있다면 4가 4개씩 묶은 카드 번호를 의미한다는 알 수 있을 것이라고 생각해서 4로 놔두었는데요. 변수 이름을 좀 더 고민하는 게 맞았던 것 같습니다!

문자열과 숫자 단위를 받는 좀 더 일반화된 함수로 놔두는 것에 대해서도 사실 고민을 하긴 했었습니다. 가장 처음 든 생각은 좀 더 일반화 된 함수를 만들었을 때, 사용하는 곳이 있는가? 하는 생각이었고 또 다른 생각은 최근에 Good Code, Bad Code 라는 개발 서적을 가지고 하는 북스터디에서 기존 코드에서 좀 더 일반화된 함수로 쪼개는 작업에 대한 주제를 다루었었는데, 코드를 분리했을 때 만약에 분리되기 전의 함수가 분리된 함수에 의존하고 있는데 분리된 함수가 노출된다면 분리된 함수를 다른 개발자가 사용한 후부터는 새로운 의존 관계 때문에 분리된 함수의 변경이 어려워 지는 문제가 발생할 수도 있다는 것을 알게되었습니다. 그래서 고민을 하다가 그냥 분리하지 않기로 했는데요, 막상 작성해주신 코드를 읽고 보니 String type을 길이 단위로 Array로 쪼개는 함수는 충분히 분리해도 좋을 것 같다는 생각이 들었습니다!

junghyeonsu replied:
> 코드를 분리했을 때 만약에 분리되기 전의 함수가 분리된 함수에 의존하고 있는데 분리된 함수가 노출된다면 분리된 함수를 다른 개발자가 사용한 후부터는 새로운 의존 관계 때문에 분리된 함수의 변경이 어려워 지는 문제가 발생할 수도 있다는 것을 알게되었습니다.

정말 좋은 인사이트네요 저도 배우고 갑니다!

저는 항상 util 스럽게 함수를 짜는 연습을 하고, 제가 오픈소스를 정말 많이 한 것은 아니지만 오픈소스를 하면서 느낀 것은 **유틸스럽게 짠 함수 혹은 컴포넌트를 내보내는 것이 어떻게 보면 오픈소스 아닐까?**.. 라는 것을 느꼈어요.

요기서 "유틸스럽다" 는 어떤 도메인에도 엮이지 않고, 프로젝트 어디서나 혹은 모든 사람들이 사용할 수 있는 함수나 컴포넌트나 그런 것들을 말해요. (제가 정의한 내용입니다 하하) 그리고 프로젝트에서 이런 함수들이나 컴포넌트들을 모으다보면 오픈소스로 빼낼 때 그냥 그 함수를 바로 내보내면 되거든요.

그리고 저는 사내에서 모노레포 환경에서 많이 작업을 하는데 저런 유틸들을 패키지로 따로 빼서

```ts
import { splitString } from "@my-utils";
```

이런식으로 패키지 import하면 뭔가 라이브러리 만드는 것 같은 느낌도 들고 되게 기분이 좋더라구요ㅋㅋㅋㅋㅋ 이건 개인취향차이입니다.

그래서 저는 항상 유틸스럽게 만드려고 노력하고, 피드백도 그런식으로 드리고 있는 것 같네요.
저의 가치관이 많이 반영된 것 같아요.
그래서 저의 얘기가 항상 맞는다는 것은 아니고 동호님께서 걸러들으셔서 듣고싶으신 것만 들으시면 됩니다ㅎㅎㅎ

junghyeonsu commented on src/Components/Card/Card.tsx:
```diff
@@ -0,0 +1,70 @@
+  return splittedNumberArr;
+};
+
+const replaceDigitsWithDot = (
+  numberArr: string[],
+  numberOfDigitsChunkToReplace: number
+): string[] => {
+  const replacedNumberArr: string[] = [];
+  for (let i = 0; i < numberArr.length; i++) {
+    if (i < numberArr.length - numberOfDigitsChunkToReplace)
+      replacedNumberArr.push(numberArr[i]);
+    else replacedNumberArr.push('····');
+  }
+  return replacedNumberArr;
+};
```
해당 함수도 도메인이 묶여있지 않은 유틸 함수로 바꾸면 좋겠네요!
그렇게 되면 좋은 것이 테스트하기가 쉬워져요.
유닛 테스트를 쉽게 쉽게 할 수 있어서 유틸함수 짜기가 되게 좋거든요.
나중에 리팩토링 하실 때 유닛 테스트 도구 (vitest, uvu) 같은 도구들 도입하셔서 한 번 작성해보시면 괜찮을 것 같아요!

psy082 replied:
이 함수는 위의 케이스와는 다르게 쪼개지 않는 것이 더 나을 수도 있을 것 같습니다!

그 이유는 제 생각에는 위의 함수는 card 컴포넌트와 함께 있기 때문에 replaceDigitsWithDot이라는 이름으로 이 함수가 하는 일이 설명될 수 있지만 외부로 노출되는 경우에는 좀 더 고민이 필요할 것 같기 때문입니다. 위의 함수는 dot으로 변환하는 작업을 입력받은 숫자만큼의 숫자 청크를 뒤에서 부터 dot으로 치환합니다. 또한 입력 받은 숫자 스트링의 길이는 4입니다. 위의 함수는 사실 이 두 사실을 전제로 동작하는데요 이 지식은 다소 도메인적인 지식이고 Card와 함께 있을 때는 전제되어 있어도 크게 문제될 것 같지는 않다고 생각했습니다. 하지만 만약에 함수가 replaceDigitsWithDot라는 이름으로 외부로 노출된다면 저는 2가지를 더 고려하게 될 것 같습니다.

1. 다른 사용자는 뒤에서 부터 dot으로 치환하지 않고 앞에서부터 치환하고 싶다면? 아니면 5개의 청크 중에 2,3,4 청크를 치환하고 싶다면? 아마 외부로 노출된 함수는 좀 더 일반화된 상황을 처리하기 위해서 수정이 필요하게 될 것 같습니다. 아마 함수 이름을 replaceDigitsWithDotFromBehind로 좀 더 구체적으로 정할 수도 있을 것 같긴 합니다.

2. 입력 받는 숫자 청크들의 숫자열 길이가 서로 다르다면? 함수 이름 만으로는 입력받는 숫자열 배열의 숫자열 길이가 4라는 사실이 명시되지 않았기 때문에 ['123', '2345', '1', '34'] 와 같은 입력을 처리할 수 있도록 내부를 수정하거나, replaceFourDigitsWithDotFromBehind과 같이 4자리 숫자만 치환한다는 것이 명시되어야 할 것 같습니다. (다만 이 경우에는 실제로 사용자가 4자리 숫자열들만 넘긴다는 것을 코드적으로 보장할 수 없을 것 같습니다... -> 생각해보니 이 부분은 Card 컴포넌트가 카드 번호를 숫자열 배열로 받을 때에도 동일한 문제가 발생하겠네요. Card 컴포넌트에서 validation을 하던지 아니면 사용자가 제대로 넘겨준다는 전제 하에 해야 할 것 같은데, Card 번호는 도메인적 지식이니 전제를 해도 괜찮지 않을까 싶기도 합니다. 음... 이 것도 고민이 되네요...)

제가 생각하기에는 replaceDigitsWithDot 분리를 시키기 위해서는 replaceDigitsWithDot의 유틸함수화를 위해 고민이 더 필요할 것 같다고 생각해서 Card에 대한 지식을 다소 전제한 상태로 Card 컴포넌트에 붙어있는 것이 더 좋지 않을까 하는 생각이 들었습니다...!

jest 이외의 도구들은 전혀 몰랐는데, vitest,와 uvu 도구도 같이 살펴보겠습니다!

junghyeonsu replied:
> 이 지식은 다소 도메인적인 지식이고 Card와 함께 있을 때는 전제되어 있어도 크게 문제될 것 같지는 않다고 생각했습니다.

넵 이렇게 생각하신다면 그대로 진행하시면 돼요! 어디까지나 전적으로 제 생각이니까요.

> 하지만 만약에 함수가 replaceDigitsWithDot라는 이름으로 외부로 노출된다면 저는 2가지를 더 고려하게 될 것 같습니다.

저는 이 부분에서 저의 가치관이 또 드러난다고 생각했는데요 하하..
이런 고민을 하는 것 자체가 "오픈소스 마인드" 라고 할까요? 항상 오픈소스로 뺄 때는 이런 것들을 더 생각하게 되더라구요.
어떤 예외상황들이 있을까? 어떻게 하면 다양한 케이스들을 충족시킬 수 있을까? 이런 고민이 저는 코딩에서 되게 좋은 고민들 인 것 같아서요.

하지만 저희 이 수업은 오픈소스 수업은 아니니, 저는 동호님의 의견도 완전히 이해하고 이대로 진행하셔도 무방해요.
단지 저는 리뷰어로서 저의 생각을 나눠드린다는 것 뿐이고 저의 의견을 사는(buying)것은 전적으로 동호님의 선택이겠네요.

1. 저는 만약 함수들을 생성한다면 `다른 사용자는 뒤에서 부터 dot으로 치환하지 않고 앞에서부터 치환하고 싶다면?` 이런 니즈가 있으면 옵션으로 열어야 되는 순간이라고 생각하기는 해요. 유틸 함수 안에서 배열 앞에서 처리를 기본적으로 하고 reverse 라는 옵션이 있어서 true로 들어오면 뒤에서 부터 처리를 해준다? 이런 간단한 생각을 해볼 수 있겠네요.

그리고 추가적으로 `dot` 으로 치환해준다도 도메인에 엮여있어서 그냥 **치환해주는 문자**도 옵션으로 받을 수 있지 않을까? 이 생각도 했어요. transformString 과 같은 이름을 주거나 조금 더 나은 이름이 있을 수 있겠네요.

2. 일단 이 Card 컴포넌트는 단순 받은 props을 통해서 UI를 그려주는 dumb한(아무런 로직이 없는) 컴포넌트라고 생각을 했구요. 만약 그렇다고 한다면 카드 번호가 잘못들어온 상황에 있어서는 함수안에서는 처리를 우선은 해주지 않을 것 같긴하네요! 이 부분은 이 전에 카드를 저장할 때 검증이 되어야 하는 상황이라고 생각했어요.

만약 검증이 되지 않은 카드 번호에 대해서 처리를 해주어야 하는 유틸 함수를 만들어야 한다고 생각한다면 저는 4 라는 길이를 제한시키지 않고 일단 배열을 받고, 배열을 순회하면서 string의 길이만큼 transformString(위에서 말한 치환해주는 문자) 로 변환시켜 주면 되지 않을까 하는 가벼운 생각을 해봤어요.

너무 가볍게 얘기해서 동호님의 고민을 다 해결을 못해드리겠지만,, 조금이나마 도움이 되었으면 좋겠네요.
제 생각을 전부 사실 필욘 없습니다ㅎㅎ

아, 그리고 vitest는 jest보다 훨씬 가볍고 아주 컴팩트한 테스트 도구로써 유틸 테스트하기에 아주 좋은 테스트 도구에요.
uvu는 그것보다 더 ~ 가벼운 테스트 도구로 진짜 유틸들만 검증할 수 있는 도구구요.
나중에 꼭 써보세요 도입하는것도 간단해서요ㅎㅎ

junghyeonsu commented on src/Components/CardInput/useComponent.tsx:
```diff
@@ -0,0 +1,105 @@
+  nextRef: RefObject<HTMLInputElement>;
+}
+
+class Components {
```
해당 Components class와 밑에 ComponentContext는 Input 컴포넌트들을 통채로 넣고 삭제하고 추가하고 가져오는 로직이 있는 것 같아요.

정확하게 이 컨텍스트가 왜 필요한 지 제가 이해를 못한 것 일수도 있는데요!
그리고 Input 컴포넌트에서는 렌더링 됐을 때 add하고 dom에서 빠질 때 remove해주고
input 변할 때 components에 담겨져 있는 것 중에 해당 Ref에 해당하는 값만 변경해주고
이런 로직들이 있는 것 같은데 그럴거면 사실 단순히 InputContext를 만들어서 input 하나마다 달아주면 되는 것 아닌가? 하는 생각을 했어요!

정확하게 useComponent의 목적과 동호님이 하고자 하는 작업이
> focus 이동 기능, 입력 값 길이 계산 기능 들이 외부에 노출되지 않고 컴포넌트 내부에서 제공되도록

와 같은 작업들이었다면 제 생각에 단순 context만으로도 해결이 될 수 있는 부분 같거든요!
그래서 혹시나 이 부분에 맥락이 더 있다면 저에게 설명해주시면 감사드려요!

psy082 replied:
현수님 말씀을 듣고 보니 제가 해결하려고 하는 문제를 너무 복잡하게 생각했던 것 같다고 느꼈습니다. 문제는 제가 

> 단순히 InputContext를 만들어서 input 하나마다 달아주면 되는 것 아닌가? 

이 부분을 해결하기 위한 단순한 방법을 생각하지 못한 것 같습니다...
 
해당 컴포넌트 클래스를 작성한 맥락을 좀 더 설명하겠습니다. 

저는 CardInput 컴포넌트 하위의 input들 중 어떤 input이든지 max length 만큼 입력되었다면 다음 input으로 focus가 넘어가기를 원했습니다. 이를 위해서는 2가지 정보가 필요했습니다. 현재 max length 정보를 확인하기 위한 현재 input dom 정보와 focus를 넘길 다음 input dom 입니다. 제가 원했던 방식은 현재 input dom 정보를 주면 다음 input dom 정보를 반환하는 것을 원했고, 이 방식을 위한 데이터 형식으로 map 형식이 적절하다고 생각했습니다. key에는 현재 input dom 정보가, value에는 다음 input dom 정보가 있는 형식으로요. 

그런데 value에는 다음 input dom 정보 뿐만 아니라 key와 동일한 현재 input dom 정보도 필요했습니다. 입력 값 길이 계산을 위해서 모든 input을 흝을 때 필요하기 때문에요. 그래서 처음에는 key에는 현재 input dom을, value에는 현재 input dom과 다음 input dom이 담긴 객체가 되도록 작성을 했습니다. 이 map에 데이터를 추가하는 방식을 클래스 메소드 내부로 감추는 것이 좋겠다고 생각해서 map과 map에 데이터를 추가하는 일을 Components 클래스에 내장했습니다. 

그리고 Input 컴포넌트에서는 dom이 생성된 이후에 dom 정보를 context에 저장되도록 useEffect 훅에서 ref를 저장하게 했습니다. *여기서 고민이 시작되었는데요* useEffect에 clear 함수를 사용해야 하는가? 였습니다. 만약에 ref에 dom을 저장했는데 컴포넌트 삭제로 인해 실제로 dom이 존재하지 않는다면 삭제된 dom을 next로 가리키고 있는 이전 dom은 nextRef에 존재하지 않는 dom 정보를 저장하고 있는 것이 되었습니다. 그래서 이전 dom이 가리키고 있는 nextRef 정보를 갱신해주어야 하지 않은가? 해서 remove 메소드를 작성하게 되었습니다. 그런데 여기서 내가 삭제되었을 때, 나를 nextRef로 가지고 있는 input dom 정보도 필요해졌기 때문에 추가적으로 prevRef 정보도 같이 담게 되었습니다.

이 정도까지 작성한 후에 드는 고민은 어쩌면 현수님의 의문과 비슷했습니다. 어차피 CardInput 내부 input 들은 CardInput과 생성과 삭제 시점이 동일할 텐데 개별 input 삭제에 대한 고려가 필요한가? 였습니다. 여기서 든 생각은 현재 작성된 Components class와 Components class 정보에 의존하고 있는 useComponentContext는 분리된 시점에서 사실상 CardInput에 대한 도메인적 지식이 없다는 것이었습니다. 만약에 Components가 CardInput과 CardInput 내부의 input들이 생성과 삭제 시점이 동일하기 때문에 개별 input의 생성과 삭제를 고려하지 않아도 된다는 지식을 가지게 되면 CardInput과 좀 더 강하게 결합하게 된다는 생각을 하게 되었습니다.(물론 useComponent가 CardInput 폴더 내부에 있다는 것이 함정이긴 합니다...) 그래서 훅으로 분리한 시점에서 좀 더 복잡하게 구현을 하게 되었습니다...

설명이 길었는데, 저의 고민에 현수님의 생각은 어떠신지 궁금합니다!

p.s. 설명을 적다보니 훅이 CardInput 폴더 내부에 존재한다면 굳이 개별 input의 생성과 삭제를 고려하지 않아도 될 것 같다는 생각도 들었습니다. 그리고 애초에 remove를 만들 고민이 시작된 시점도 useEffect에서 clear 함수를 만들어야 하는가? 였는데 애초에 그 시점에서 clear 함수가 불필요할 수도 있었겠네요. 이건 결론적으로 useComponent hook을 어디까지 노출시킬 것인가의 문제인 것인가... 하는 생각이 들기도 하네용...

junghyeonsu replied:
동호님 제가 이거 고민하느라 정말 답변이 많이 늦어졌네요.

저도 직접 한 번 구현을 해봤는데요. 저는 동호님이 구현해놓으신 부분이 custom hook 같다고 생각해서 커스텀 훅 처럼 구현을 해봤어요.

우선 코드를 먼저 보여드릴게요.

### useInput.ts

```ts
import type { RefObject } from "react";
import { useState } from "react";
import { useEffect } from "react";

interface Value {
  name: string;
  ref: RefObject<HTMLInputElement>;
  value: string;
}

interface UseInputProps {
  values: Value[];
  maxLength: number;
}

export const useInput = ({ maxLength, values }: UseInputProps) => {
  const [inputValues, setInputValues] = useState(values);
  const [order, setOrder] = useState(0);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target as HTMLInputElement;
    const newValues = inputValues.map((inputValue) => {
      if (inputValue.name === event.target.name) {
        return {
          ...inputValue,
          value,
        };
      }
      return inputValue;
    });

    setInputValues(newValues);

    if (value.length === maxLength) {
      const next = order + 1;
      if (next < inputValues.length) {
        setOrder(next);
      }
    }
  };

  useEffect(() => {
    const input = inputValues[order];
    if (input) {
      input.ref.current?.focus();
    }
  }, [order]);

  return { handleChange, inputValues };
};

```

### 그리고 사용할 때 (App.tsx)

```tsx
import "./App.css";

import { useRef } from "react";

import { useInput } from "./useInput";

function App() {
  const firstInput = useRef<HTMLInputElement>(null);
  const secondInput = useRef<HTMLInputElement>(null);
  const thirdInput = useRef<HTMLInputElement>(null);
  const fourthInput = useRef<HTMLInputElement>(null);

  const { handleChange, inputValues } = useInput({
    values: [
      {
        name: "first",
        ref: firstInput,
        value: "",
      },
      {
        name: "second",
        ref: secondInput,
        value: "",
      },
      {
        name: "third",
        ref: thirdInput,
        value: "",
      },
      {
        name: "fourth",
        ref: fourthInput,
        value: "",
      },
    ],
    maxLength: 4,
  });

  return (
    <div className="App">
      <input
        value={inputValues[0].value}
        onChange={handleChange}
        ref={firstInput}
        type="text"
        name="first"
      />
      <input
        value={inputValues[1].value}
        onChange={handleChange}
        ref={secondInput}
        type="text"
        name="second"
      />
      <input
        value={inputValues[2].value}
        onChange={handleChange}
        ref={thirdInput}
        type="text"
        name="third"
      />
      <input
        value={inputValues[3].value}
        onChange={handleChange}
        ref={fourthInput}
        type="text"
        name="fourth"
      />
    </div>
  );
}

export default App;

```

확실히 자동으로 ref 넘기고 하는 이런 로직들이 많이 보이는 패턴은 아니라서 저도 구현을 하는데 꽤 고민을 많이했던 것 같네요.
제가 작성한 코드들이 정답은 아니고 고쳐야 할 부분도 꽤 보이네요.

이거 제가 슬랙에 다 같이 볼 수 있도록 더 상세하게 남겨볼게요

junghyeonsu reviewed (changes requested):
안녕하세요 동호님~
많은 고민이 있으셨군요. 하나씩 얘기해볼게요!
페이지 개발이 다 되시면 다시 또 리뷰 요청주세요!

1.
지금 index.tsx에서 App.tsx를 부르고있는데 App에는 아무것도 없는 것 같아서 이거 데모가 안될 것 같은데? 하는 생각을 했어요. 현재 진입점이 없는게 맞겠죠!?

2.
> 처음에는 기능만 생각하고 컴포넌트 구현을 하다가 나중에는 스타일 관련 정보도 컴포넌트 사용자에게 입력 받아야 한다는 사실을 깨달았습니다. 지금은 스타일을 전부 외부에서 입력 받도록 노출하지는 않았고 width를 위한 %값과 background 사용 여부만 노출했는데, 개발 과정에서 기능보다는 스타일 때문에 컴포넌트의 props가 자주 변경될 여지가 많을 것 같다는 것을 느꼈습니다. 리액트에서는 스타일 관련 변화율을 어떻게 관리하는 지 궁금합니다! 아예 외부에서 style을 지정할 수 있도록 style을 props로 노출하는 것이 더 나을까요?

이건 동호님이 생각하시는 디자인시스템이 어떤 컨셉을 가지고 있느냐에 따라서도 다르고, 동호님이 진행하시는 프로젝트의 성격에 따라서도 다른 것 같아요. 디자인시스템에도 그 회사의 특색이 많이 섞여있는 컴포넌트들이 많은 디자인시스템도 있구요, 범용적으로 사용되기 위한 디자인시스템도 있어요. 만약 동호님께서 생각하시는 프로젝트의 디자인시스템이 **유저들에게 많은 선택지를 열어주는 컴포넌트를 제공한다** 라는 것을 목표로 하신다면 props를 많이 열어둘 수도 있어요. 그게 아니라 **사용자들이 우리가 원하는 방식으로만 컴포넌트를 사용했음 좋겠다.** 하고 생각한다면 props를 제한적으로 두고 유저들이 커스터마이징 못하도록 막아둘 수도 있겠죠. 이건 선택이에요.

그리고 현재 width를 %로 받고 계신 것 같아요! 제가 처음 봤을 때는 저게 px인지 rem인지 %인지 헷갈렸거든요. 만약 컴포넌트를 모두가 사용한다면 그냥 string으로 width를 받고 "100%" 혹은 "300px" 이런식으로 받을 수 있는 것이 훨씬 더 명확한 것 같아요. 유저입장에서는 해당 옵션이 조금 헷갈릴 수도 있겠네요.

3.
> 지금의 컴포넌트가 css class 이름에 props가 강하게 의존하고 있는 것도 문제로 느끼는 데요, 어쩌면 css in js가 그 모든 정보를 컴포넌트 내부로 감추면서 문제를 해결할 수도 있을 것 같지만, 최근에는 css in js를 사용하지 않다는 글도 읽었던 것 같아서요. 컴포넌트에서 스타일 정보와의 의존 문제를 어떤 관점에서 보는 게 좋을 지도 궁금합니다. 저는 아직 무엇을 기준으로 판단하는 게 좋을지 모르겠습니다...

**최근에는 css in js를 사용하지 않다는 글도 읽었던 것 같아서요** 와 같은 글에서 혹시 어떤 이유로 사용하지 않는다고 했던 걸까요? 사실 스타일링도 목적에 따라 프로젝트 성격에 따라 다 다른 방식으로 사용할 수 있는 것 같아요. 현재는 native css를 그대로 쓰셨군요. 근데 쓰시면서 className 지정과 같은 이슈가 있고 그게 크게 느껴지신다면 저는 css in js 기법도 나쁘지 않을 것 같네요. css in js가 요즘 유행하는 Next 프레임워크나 SSR 환경에서 대응이 조금 어렵다는 점도 있고, 결국은 JS라서 style로 트랜스폼 하는 과정도 필요해서 native  css보다는 성능이 조금 떨어진다는 평도 있구요. 하지만 지금과 같은 애니메이션이 많이 들어가는 프로젝트도 아니고 성능이 엄청 중요한 프로젝트는 아니라서 css-in-js는 괜찮은 선택지라고 저는 생각들어요.

https://vanilla-extract.style/

제가 아는 css-in-js 중에 바닐라 익스트랙트 라는 라이브러리인데 빌드 타임에 native css로 변환해주는 라이브러리에요. 그리고 css를 아토믹하게 조합해서 쓸 수도 있고, 타입스크립트도 잘 지원해줘서 저는 되게 괜찮게 썼던 기억이 있네요. 이런 것들도 한 번 알아보시면 좋을 것 같아요!

junghyeonsu reviewed (changes requested):
우선 동호님 고생하셨어요!
step1 미션 우선 다 하시고 저한테 리뷰주시면 빠르게 머지하고, 스텝 투로 넘어가시죠!

psy082 committed:
chore: package.json 의존성 수정

- 일부 의존성 devDependencies로 이동
- import 순서 포매팅을 위한 플러그인 추가
- util 함수 unit test를 위한 테스트 도구, uvu 추가

psy082 committed:
refactor: Card Component Props 수정

- 카드 번호 실시간 반영을 위해 카드 번호 형식을 string -> string[]으로 변경
- 카드 번호를 *으로 변환하는 함수를 유틸 함수로 분리

psy082 committed:
refactor: CardInput Container width의 type을 템플릿 문자열 타입에서 string으로 변경

기존 width의 템플릿 문자열 타입이 px단위인지 %단위인지 모호하고
제공되는 % 값이 제한적이므로 앞으로 변경될 가능성이 높기 때문에 받는 타입을 문자열로 변경

psy082 committed:
feat: Modal Component 구현

- ref로 wrapping된 parent dom 하위에 portal이 생성되도록 구현
- dimmed container 클릭 시 모달이 닫히는 기능 내장

psy082 committed:
feat: Card Form 입력 페이지 작성

- 작성된 CardInput 컴포넌트를 활용하기만 한 구현
툴팁, 자동 포커스, 비밀 키보드 기능 등을 추가하기 위해서
CardInput 컴포넌트 수정 후 추가로 작업이 필요함

psy082 committed:
add: CardInput Container 기능 추가

- 입력되지 않은 input으로 자동 포커싱 되는 기능 추가 및 해당 기능 ref를 통해 외부에 노출
- validate 기능도 ref를 통해 외부에 노출

psy082 committed:
feat: 카드 회사 선택을 위한 모달 기능 구현

- 앞의 8자리만 입력된 경우에 카드회사 선택 모달이 생성됨
- 앞의 8자리가 카드 BIN 번호와 일치할 경우 바로 회사가 선택됨

모달 기능 구현으로 인해 추가 및 변경된 사항

- 기존에는 입력이 완료되지 않은 input으로 auto focus 되었으나
  값이 하나도 입력되지 않은 경우에만 제일 앞으로 auto focus되도록 수정
- backspace 입력 시 input에 값이 없는 경우에는 앞의 input이 focus되는 기능 추가

psy082 committed:
feat: 다음 버튼 클릭 시 form validation 되는 기능 구현

- CardInput에서 useInperativeHandle을 통해 외부로 validation 제어권 노출
- 다음 버튼 클릭 시 CardInput에서 노출한 validation을 실행

psy082 commented:
현수님 안녕하세요!
지난 번에 리뷰해주신 내용을 반영했습니다!
카드 입력 페이지 요구 사항은 전부 구현을 한 것 같습니다. 

하지만 아직 router를 붙이지 못했고
(그래서 카드 입력 페이지도 아직까지는 개발 환경에서 확인은 안되고 스토리북에서만 확인이 가능합니다...)

카드 목록 페이지와 카드 등록 완료 페이지 작성을 못해서  
아래의 두 기능은 아직 완료되지 못한 상황입니다.

- (뒤로가기) 버튼 클릭 시, 카드 목록 페이지로 이동한다.
- 카드 추가 완료시 카드 등록 완료 페이지로 이동한다.

위의 두 기능은 카드 목록 페이지와 카드 등록 완료 페이지를 만들어야 하니 step 2에서 해야 되는 건가 갸우뚱하네요...

지난번 리뷰해 주신 내용 중에 카드 번호 청크를 특정 문자로 치환하는 함수를 유틸함수화 하면 좋겠다고 하신 내용이 있었는데요. 
기능을 구현하다보니 해당 기능은 카드 컴포넌트가 카드 번호를 string[]로 받으면서 도메인적 지식을 덜어내고 유틸함수화하는 것이 낫다고 판단해서 유틸함수로 구현해보았습니다. 
소개해주신 도구 중에 uvu를 선택해서 도입해 보았었는데요. uvu 개발자의 의도인 것 같긴 한데, 최소한의 기능만 남겨두기 위해서 watch 기능이 없기도 하고(저는 watch 기능을 켜고 구현하는 편이라 없는 게 불편하더라고용) uvu 사용을 위해서는 esm, tsm과 같은 문법 해석 도구가 추가로 설치되어야 해서 기본적으로 제공되는 jest로 테스트 도구를 바꾸었습니다. vitest는 uvu에 시간을 좀 쏟았어서 도입해 보지는 못했는데, 다음 기회에 사용해보려구용. 감사합니다 :) 

css를 적절하게 쪼개거나 selector 이름을 정하는 부분은 충분히 고민하지 못했어서 엉성한 상태 입니다.
css에 대한 고민은 아직 많이 필요한 상태인데 장바구니 미션 전에 고민을 더 할 수 있을지 잘 모르겠습니다...
(고민을 더 해야 되는데 말이죠...)
